tf_quant_finance/black_scholes/approximations/american_option.py

```diff
@@ -0,0 +1,290 @@
+                  x,
+                  t,
+                  r,
+                  b,
+                  s=None,
+                  is_call_options=None,
+                  dtype=None,
+                  max_iterations=20,
+                  tolerance=1e-8,
+                  name=None):
+  """Computes prices of American options using the Baron-Adesi Whaley approximation."""
+
+  # The naming convention will align variables with the variables named in
+  # reference [1], but made lower case, and differentiating between put and
+  # call option values with the suffix _put and _call.
+  # [1] https://deriscope.com/docs/Barone_Adesi_Whaley_1987.pdf
+
+  with tf.name_scope(name):
+    where_call = tf.broadcast_to(
+      tf.constant(True), s.shape) if is_call_options is None else is_call_options
+
+    sign = tf.where(
+      where_call, tf.constant(1, dtype=dtype), tf.constant(-1, dtype=dtype))
+
+    q2, a2, s_crit, converged, failed = _option_price_formula_variables(
+      sigma, x, t, r, b, sign,
+      is_call_options, dtype, max_iterations=max_iterations, tolerance=tolerance)
+
+    eu_prices = vanilla_prices(
+      volatilities=sigma, strikes=x, expiries=t, spots=s, discount_rates=r,
+      cost_of_carries=b, is_call_options=is_call_options, dtype=dtype,
+      name=name)
+
+    # The divisive condition is different for put and call options
+    condition = tf.where(where_call, s < s_crit, s > s_crit)
+
+    american_prices = tf.where(
+      condition,
+      eu_prices + a2 * (s / s_crit) ** q2,
+      (s - x) * sign)
+
+    return american_prices, converged, failed
+
+
+def _option_price_formula_variables(sigma,
+                                    x,
+                                    t,
+                                    r,
+                                    b,
+                                    sign,
+                                    is_call_options,
+                                    dtype,
+                                    max_iterations=20,
+                                    tolerance=1e-8):
+  """Computes prices of American options using the Baron-Adesi Whaley approximation."""
+
+  # The naming convention will align variables with the variables named in
+  # reference [1], but made lower case, and differentiating between put and
+  # call option values with the suffix _put and _call.
+  # [1] https://deriscope.com/docs/Barone_Adesi_Whaley_1987.pdf
+
+  m = 2 * r / sigma ** 2
+  n = 2 * b / sigma ** 2
+  k = 1 - tf.exp(- r * t)
+  q_ = _calc_q(n, m, sign, k)
+
+  value = lambda s_crit: vanilla_prices(
+    volatilities=sigma, strikes=x, expiries=t, spots=s_crit, discount_rates=r,
+    cost_of_carries=b, is_call_options=is_call_options, dtype=dtype) + sign * (
+      1 - tf.exp((b - r) * t) * _ncdf(sign * _calc_d1(
+    s_crit, x, sigma, b, t))) * s_crit / q_ - sign * (s_crit - x)
+
+  value_and_gradient_func = lambda price: value_and_gradient(value, price)
+
+  # Calculate seed value for critical spot price
```

Comment:
seed is usually have something to do with random numbers. Could you please point to the paper explicitly ?